incubator/hnc/pkg/reconcilers/hierarchical_namespace.go

```diff
@@ -31,44 +39,176 @@ type HierarchicalNamespaceReconciler struct {
 
 	forest *forest.Forest
 	hcr    *HierarchyConfigReconciler
+
+	// Affected is a channel of event.GenericEvent (see "Watching Channels" in
+	// https://book-v1.book.kubebuilder.io/beyond_basics/controller_watches.html) that is used to
+	// enqueue additional objects that need updating.
+	Affected chan event.GenericEvent
 }
 
 // Reconcile sets up some basic variables and then calls the business logic. It currently
 // only handles the creation of the namespaces but no deletion or state reporting yet.
 func (r *HierarchicalNamespaceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	// TODO report error state if the webhook is bypassed - see issue
-	//  https://github.com/kubernetes-sigs/multi-tenancy/issues/459
-	if ex[req.Namespace] {
-		return ctrl.Result{}, nil
-	}
-
+	ctx := context.Background()
 	log := r.Log.WithValues("trigger", req.NamespacedName)
 	log.Info("Reconciling HNS")
 
-	// Names of the hierarchical namespace and the current namespace.
+	// Get names of the hierarchical namespace and the current namespace.
 	nm := req.Name
 	pnm := req.Namespace
 
-	// Set RequiredChildOf (Owner) in the forest of the hierarchical namespace to the
-	// current namespace.
-	// TODO rename RequiredChildOf to Owner in the forest. See issue:
-	//  https://github.com/kubernetes-sigs/multi-tenancy/issues/469
-	r.forest.Get(nm).RequiredChildOf = pnm
+	// Get instance from apiserver. Even though the instance doesn't have spec, we don't need to save
+	// this api call since otherwise an api read is required before writeInstance() anyway.
+	inst, err := r.getInstance(ctx, pnm, nm)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Report "Forbidden" state and early exist if the namespace is not allowed to self-serve
+	// namespaces but has bypassed the webhook and successfully created the hns instance.
+	// TODO refactor/split the EX map for 1) reconciler exclusion and 2) self-serve not allowed
+	//  purposes. See issue: https://github.com/kubernetes-sigs/multi-tenancy/issues/495
+	if EX[pnm] {
+		inst.Status.State = api.Forbidden
+		err := r.writeInstance(ctx, log, inst)
+		return ctrl.Result{}, err
+	}
+
+	// Get the self-serve namespace's hierarchyConfig and namespace instances.
+	hcInst, nsInst, err := r.hcr.GetInstances(ctx, log, nm)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.syncWithForest(log, inst, nsInst, hcInst)
+
+	// TODO report the "RequiredChildConflict" in the HNS reconciliation. See issue:
```

Comment:
We should change `RequiredChildConflict` to a better name, eg `SubnamespaceConflict` or something like that. We don't use the term "Required Child" in the new UX.